File: js/src/Vista.SDK/LocalId.Items.ts
import { GmodPath } from '.';

export class LocalIdItems {
  public primaryItem?: GmodPath;
  public secondaryItem?: GmodPath;

  public append(builder: string[], verboseMode?: boolean) {
    if (!this.primaryItem && !this.secondaryItem) return;

    if (!!this.primaryItem) {
      builder.push(this.primaryItem.toString());
      builder.push('/');
    }

    if (!!this.secondaryItem) {
      builder.push('sec/');
      builder.push(this.secondaryItem.toString());
      builder.push('/');
    }

    if (verboseMode) {
      if (!!this.primaryItem) {
        for (const { depth, name } of this.primaryItem.getCommonNames()) {
          builder.push('~');
          const location = this.primaryItem.getNode(depth).location;
          LocalIdItems.appendCommonName(builder, name, location);
          builder.push('/');
        }
      }

      if (!!this.secondaryItem) {
        let prefix = '~for.';
        for (const { depth, name } of this.secondaryItem.getCommonNames()) {
          builder.push(prefix);
          if (prefix !== '~') prefix = '~';
          const location = this.secondaryItem.getNode(depth).location;
          LocalIdItems.appendCommonName(builder, name, location);
          builder.push('/');
        }
      }
    }
  }
  private static appendCommonName(builder: string[], commonName: string, location?: string) {
    let prev: string | undefined = undefined;
    for (const ch of commonName) {
      if (ch === '/') continue;
      if (prev === ' ' && ch === ' ') continue;

      switch (ch) {
        case ' ':
          builder.push('.');
          break;
        default:
          builder.push(ch.toLocaleLowerCase());
          break;
      }
      prev = ch;
    }

    if (!!location) {
      builder.push('.');
      builder.push(location);
    }
  }
}

File: csharp/src/Vista.SDK/LocalIdItems.cs
using System.Text;

namespace Vista.SDK;

internal readonly record struct LocalIdItems
{
    public GmodPath? PrimaryItem { get; init; }

    public GmodPath? SecondaryItem { get; init; }

    internal void Append(StringBuilder builder, bool verboseMode)
    {
        if (PrimaryItem is null && SecondaryItem is null)
            return;

        if (PrimaryItem is not null)
        {
            PrimaryItem.ToString(builder);
            builder.Append('/');
        }

        if (SecondaryItem is not null)
        {
            builder.Append("sec/");
            SecondaryItem.ToString(builder);
            builder.Append('/');
        }

        if (verboseMode)
        {
            if (PrimaryItem is not null)
            {
                foreach (var (depth, name) in PrimaryItem.GetCommonNames())
                {
                    builder.Append('~');
                    var location = PrimaryItem[depth].Location;
                    AppendCommonName(builder, name, location?.ToString());
                    builder.Append('/');
                }
            }

            if (SecondaryItem is not null)
            {
                var prefix = "~for.";
                foreach (var (depth, name) in SecondaryItem.GetCommonNames())
                {
                    builder.Append(prefix);
                    if (prefix != "~")
                        prefix = "~";

                    var location = SecondaryItem[depth].Location;
                    AppendCommonName(builder, name, location?.ToString());
                    builder.Append('/');
                }
            }
        }

        static void AppendCommonName(StringBuilder builder, string commonName, string? location)
        {
            char? prev = null;
            foreach (ref readonly var ch in commonName.AsSpan())
            {
                if (ch == '/')
                    continue;
                if (prev == ' ' && ch == ' ')
                    continue;

                var current = ch;
                switch (ch)
                {
                    case ' ':
                        current = '.';
                        break;
                    default:
                        var match = VIS.MatchISOString(ch);
                        if (!match)
                        {
                            current = '.';
                            break;
                        }
                        current = char.ToLower(ch);
                        break;
                }
                if (current == '.' && prev == '.')
                    continue;
                builder.Append(current);
                prev = current;
            }

            if (location is { Length: > 0 })
            {
                builder.Append('.');
                builder.Append(location);
            }
        }
    }
}
